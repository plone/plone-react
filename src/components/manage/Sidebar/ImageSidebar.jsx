import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Form } from 'semantic-ui-react';
import { Grid, Segment } from 'semantic-ui-react';
import { Icon, TextWidget } from '@plone/volto/components';
import { AlignTile, flattenToAppURL } from '@plone/volto/helpers';
import { settings } from '~/config';

import folderSVG from '@plone/volto/icons/folder.svg';
import imageSVG from '@plone/volto/icons/image.svg';
import clearSVG from '@plone/volto/icons/clear.svg';

const ImageSidebar = ({
  data,
  tile,
  onChangeTile,
  openObjectBrowser,
  required = false,
}) => {
  const [alt, setAlt] = useState(data.alt || '');

  return (
    <Segment.Group raised>
      <header className="header pulled">
        <h2>Image</h2>
      </header>

      {!data.url && (
        <>
          <Segment className="sidebar-metadata-container" secondary>
            No image selected
            <Icon name={imageSVG} size="100px" color="#b8c6c8" />
          </Segment>
        </>
      )}
      {data.url && (
        <>
          <Segment className="sidebar-metadata-container" secondary>
            {/* {props.data.url && <div>{props.data.url}</div>} */}
            {data.url.split('/').slice(-1)[0]}
            {data.url.includes(settings.apiPath) && (
              <img
                src={`${flattenToAppURL(data.url)}/@@images/image/mini`}
                alt={alt}
              />
            )}
            {!data.url.includes(settings.apiPath) && (
              <img src={data.url} alt={alt} style={{ width: '50%' }} />
            )}
          </Segment>
          <Segment className="form sidebar-image-data">
            {data.url.includes(settings.apiPath) && (
              <TextWidget
                id="Origin"
                title="Origin"
                required={false}
                value={data.url.split('/').slice(-1)[0]}
                icon={folderSVG}
                iconAction={() => openObjectBrowser()}
                onChange={() => {}}
              />
            )}
            {!data.url.includes(settings.apiPath) && (
              <TextWidget
                id="External"
                title="External URL"
                required={false}
                value={data.url}
                icon={clearSVG}
                iconAction={() =>
                  onChangeTile(tile, {
                    ...data,
                    url: '',
                  })
                }
                onChange={() => {}}
              />
            )}
            <TextWidget
              id="alt"
              title="alt"
              required={false}
              value={alt}
              onChange={(name, value) => {
                onChangeTile(tile, {
                  ...data,
                  alt: value,
                });
                setAlt(value);
              }}
            />
            <Form.Field inline required={required}>
              <Grid>
                <Grid.Row>
                  <Grid.Column width="4">
                    <div className="wrapper">
                      <label htmlFor="field-align">align</label>
                    </div>
                  </Grid.Column>
                  <Grid.Column width="8" className="align-tools">
                    <AlignTile
                      align={data.align}
                      onChangeTile={onChangeTile}
                      data={data}
                      tile={tile}
                    />
                  </Grid.Column>
                </Grid.Row>
              </Grid>
            </Form.Field>
          </Segment>
        </>
      )}
    </Segment.Group>
  );
};

ImageSidebar.propTypes = {
  data: PropTypes.objectOf(PropTypes.any).isRequired,
  tile: PropTypes.string.isRequired,
  onChangeTile: PropTypes.func.isRequired,
};

export default ImageSidebar;
