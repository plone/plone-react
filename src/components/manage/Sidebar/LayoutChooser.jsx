import React, { useState } from 'react';
import { Menu, Form, List, Grid } from 'semantic-ui-react';
import AlignTile from '../../../helpers/AlignTile';
import { Icon } from '@plone/volto/components';
import fitSVG from '@plone/volto/icons/image-fit.svg';
import leftSVG from '@plone/volto/icons/image-left.svg';
import rightSVG from '@plone/volto/icons/image-right.svg';
import wideSVG from '@plone/volto/icons/image-wide.svg';
import fullSVG from '@plone/volto/icons/image-full.svg';

const LayoutChooser = props => {
  const [activeItem, setActiveItem] = useState('');

  const handleChangeInput = event => {
    setActiveItem(event.target.value);
  };

  return (
    <Form.Field
      inline
      required={props.required}
      className={props.description ? 'help' : ''}
    >
      <Grid>
        <Grid.Row stretched>
          <Grid.Column width="4">
            <div className="wrapper">
              <label htmlFor={`field-${props.id}`}>{props.title}</label>
            </div>
          </Grid.Column>
          <Grid.Column width="8">
            <AlignTile
              align={props.data.align}
              onChangeTile={props.onChangeTile}
              data={props.data}
              tile={props.tile}
            />
          </Grid.Column>
        </Grid.Row>
      </Grid>
    </Form.Field>
  );
};

export default LayoutChooser;
