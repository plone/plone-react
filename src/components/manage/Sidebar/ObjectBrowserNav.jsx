import React from 'react';
import { Button, Segment } from 'semantic-ui-react';
import cx from 'classnames';
import { Icon } from '@plone/volto/components';
import { settings } from '~/config';
import rightArrowSVG from '@plone/volto/icons/right-key.svg';

const ObjectBrowserNav = ({
  currentSearchResults,
  selected,
  getIcon,
  handleClickOnItem,
  handleDoubleClickOnItem,
  mode,
  navigateTo,
}) => {
  return (
    <Segment as="ul" className="object-listing">
      {currentSearchResults &&
        currentSearchResults.items.map(item => (
          <li
            role="presentation"
            key={item.id}
            className={cx('', {
              'selected-item': selected === item['@id'],
              disabled:
                mode === 'image'
                  ? !settings.imageObjects.includes(item['@type']) &&
                    !item.is_folderish
                  : false,
            })}
            onClick={() => handleClickOnItem(item)}
            onDoubleClick={() => handleDoubleClickOnItem(item)}
          >
            <span>
              {getIcon(item['@type'])}
              {item.id}
            </span>
            {item.is_folderish && mode === 'image' && (
              <Icon name={rightArrowSVG} size="24px" />
            )}
            {item.is_folderish && mode === 'link' && (
              <Button.Group>
                <Button
                  basic
                  icon
                  onClick={e => {
                    e.stopPropagation();
                    navigateTo(item['@id']);
                  }}
                >
                  <Icon name={rightArrowSVG} size="24px" />
                </Button>
              </Button.Group>
            )}
          </li>
        ))}
    </Segment>
  );
};

export default ObjectBrowserNav;
