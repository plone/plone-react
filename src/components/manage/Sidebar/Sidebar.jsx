/**
 * Sidebar component.
 * @module components/manage/Sidebar/Sidebar
 */

import React, { Component, Fragment } from 'react';
import { Button, Tab } from 'semantic-ui-react';
import { connect } from 'react-redux';
import { compose } from 'redux';
import cookie from 'react-cookie';
import { defineMessages, injectIntl } from 'react-intl';
import cx from 'classnames';
import { BodyClass } from '../../../helpers';
import { Icon } from '../../../components';
import forbiddenSVG from '../../../icons/forbidden.svg';
import { setSidebarTab } from '../../../actions';

const messages = defineMessages({
  document: {
    id: 'Document',
    defaultMessage: 'Document',
  },
  block: {
    id: 'Block',
    defaultMessage: 'Block',
  },
  shrinkSidebar: {
    id: 'Shrink sidebar',
    defaultMessage: 'Shrink sidebar',
  },
  expandSidebar: {
    id: 'Expand sidebar',
    defaultMessage: 'Expand sidebar',
  },
});

/**
 * Sidebar container class.
 * @class Sidebar
 * @extends Component
 */
class Sidebar extends Component {
  /**
   * Property types.
   * @property {Object} propTypes Property types.
   * @static
   */
  static propTypes = {};

  /**
   * Constructor
   * @method constructor
   * @param {Object} props Component properties
   * @constructs Sidebar
   */
  constructor(props) {
    super(props);
    this.onToggleExpanded = this.onToggleExpanded.bind(this);
    this.onTabChange = this.onTabChange.bind(this);
    this.state = {
      expanded: cookie.load('sidebar_expanded') !== 'false',
    };
  }

  /**
   * On toggle expanded handler
   * @method onToggleExpanded
   * @returns {undefined}
   */
  onToggleExpanded() {
    cookie.save('sidebar_expanded', !this.state.expanded, {
      expires: new Date((2 ** 31 - 1) * 1000),
      path: '/',
    });
    this.setState({
      expanded: !this.state.expanded,
    });
  }

  /**
   * On tab change
   * @method onTabChange
   * @param {Object} event Event object
   * @param {Object} data Data object
   * @returns {undefined}
   */
  onTabChange(event, data) {
    this.props.setSidebarTab(data.activeIndex);
  }

  /**
   * Render method.
   * @method render
   * @returns {string} Markup for the component.
   */
  render() {
    const { expanded } = this.state;

    return (
      <Fragment>
        <BodyClass
          className={expanded ? 'has-sidebar' : 'has-sidebar-collapsed'}
        />
        <div className={cx('sidebar-container', { collapsed: !expanded })}>
          <Button
            aria-label={
              expanded
                ? this.props.intl.formatMessage(messages.shrinkSidebar)
                : this.props.intl.formatMessage(messages.expandSidebar)
            }
            className={
              this.props.content && this.props.content.review_state
                ? `${this.props.content.review_state} trigger`
                : 'trigger'
            }
            onClick={this.onToggleExpanded}
          />
          <Tab
            menu={{
              secondary: true,
              pointing: true,
              attached: true,
              tabular: true,
              className: 'formtabs',
            }}
            className="tabs-wrapper"
            renderActiveOnly={false}
            activeIndex={this.props.tab}
            onTabChange={this.onTabChange}
            panes={[
              {
                menuItem: this.props.intl.formatMessage(messages.document),
                pane: (
                  <Tab.Pane
                    key="metadata"
                    className="tab-wrapper"
                    id="sidebar-metadata"
                  />
                ),
              },
              {
                menuItem: this.props.intl.formatMessage(messages.block),
                pane: (
                  <Tab.Pane
                    key="properties"
                    className="tab-wrapper"
                    id="sidebar-properties"
                  >
                    <Icon
                      className="tab-forbidden"
                      name={forbiddenSVG}
                      size="48px"
                    />
                  </Tab.Pane>
                ),
              },
            ]}
          />
        </div>
        <div className={this.state.expanded ? 'pusher expanded' : 'pusher'} />
      </Fragment>
    );
  }
}

export default compose(
  injectIntl,
  connect(
    state => ({
      tab: state.sidebar.tab,
    }),
    { setSidebarTab },
  ),
)(Sidebar);
