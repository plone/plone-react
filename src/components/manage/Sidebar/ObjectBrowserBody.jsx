import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import { compose } from 'redux';
import { connect } from 'react-redux';
import {
  defineMessages,
  FormattedMessage,
  injectIntl,
  intlShape,
} from 'react-intl';
import { Input, Segment } from 'semantic-ui-react';
import { join } from 'lodash';
import { searchContent } from '@plone/volto/actions';
import { Icon } from '@plone/volto/components';
import { doesNodeContainClick } from 'semantic-ui-react/dist/commonjs/lib';

import { settings } from '~/config';
import backSVG from '@plone/volto/icons/back.svg';
import pageSVG from '@plone/volto/icons/page.svg';
import folderSVG from '@plone/volto/icons/folder.svg';
import clearSVG from '@plone/volto/icons/clear.svg';
import searchSVG from '@plone/volto/icons/zoom.svg';
import linkSVG from '@plone/volto/icons/link.svg';
import imageSVG from '@plone/volto/icons/image.svg';

import ObjectBrowserNav from './ObjectBrowserNav';

const messages = defineMessages({
  SearchInputPlaceholder: {
    id: 'Search content',
    defaultMessage: 'Search content',
  },
  ChooseImage: {
    id: 'Choose Image',
    defaultMessage: 'Choose Image',
  },
  ChooseTargetLink: {
    id: 'Choose Target',
    defaultMessage: 'Choose Target',
  },
});

function getParentURL(url) {
  return (
    `${join(url.split('/').slice(0, -1), '/')}`.replace(settings.apiPath, '') ||
    '/'
  );
}

/**
 * ObjectBrowserBody container class.
 * @class ObjectBrowserBody
 * @extends Component
 */
class ObjectBrowserBody extends Component {
  /**
   * Property types.
   * @property {Object} propTypes Property types.
   * @static
   */
  static propTypes = {
    tile: PropTypes.string.isRequired,
    mode: PropTypes.string.isRequired,
    data: PropTypes.objectOf(PropTypes.any).isRequired,
    searchSubrequests: PropTypes.objectOf(PropTypes.any).isRequired,
    searchContent: PropTypes.func.isRequired,
    closeObjectBrowser: PropTypes.func.isRequired,
    onChangeTile: PropTypes.func.isRequired,
    intl: intlShape.isRequired,
  };

  /**
   * Default properties.
   * @property {Object} defaultProps Default properties.
   * @static
   */
  static defaultProps = {
    image: '',
    href: '',
  };

  /**
   * Constructor
   * @method constructor
   * @param {Object} props Component properties
   * @constructs WysiwygEditor
   */
  constructor(props) {
    super(props);
    this.state = {
      currentFolder: this.props.data.url
        ? getParentURL(this.props.data.url)
        : '/',
      currentImageFolder: this.props.data.url
        ? getParentURL(this.props.data.url)
        : '/',
      currentLinkFolder: this.props.data.href
        ? getParentURL(this.props.data.href)
        : '/',
      parentFolder: '',
      selectedImage: this.props.data.url
        ? this.props.data.url.replace(settings.apiPath, '')
        : '',
      selectedHref: this.props.data.href
        ? this.props.data.href.replace(settings.apiPath, '')
        : '',
      showSearchInput: false,
    };
  }

  /**
   * Component did mount
   * @method componentDidMount
   * @returns {undefined}
   */
  componentDidMount() {
    document.addEventListener('mousedown', this.handleClickOutside, false);
    this.initialSearch(this.props.mode);
  }

  /**
   * Component will receive props
   * @method componentWillUnmount
   * @returns {undefined}
   */
  componentWillUnmount() {
    document.removeEventListener('mousedown', this.handleClickOutside, false);
  }

  initialSearch = mode => {
    const currentSelected =
      mode === 'image' ? this.state.selectedImage : this.state.selectedHref;
    if (currentSelected) {
      this.props.searchContent(
        getParentURL(currentSelected),
        {
          'path.depth': 1,
          sort_on: 'getObjPositionInParent',
          metadata_fields: '_all',
        },
        `${this.props.tile}-${mode}`,
      );
    } else {
      this.props.searchContent(
        '/',
        {
          'path.depth': 1,
          sort_on: 'getObjPositionInParent',
          metadata_fields: '_all',
        },
        `${this.props.tile}-${mode}`,
      );
    }
  };

  onChangeField = (name, value) => {
    this.setState({ [name]: value });
    this.onChangeTileData(name, value);
  };

  getIcon = icon => {
    switch (icon) {
      case 'Folder':
        return <Icon name={folderSVG} size="24px" />;
      case 'Document':
        return <Icon name={pageSVG} size="24px" />;
      case 'Image':
        return <Icon name={imageSVG} size="24px" />;
      case 'File':
        return <Icon name={pageSVG} size="24px" />;
      default:
        return <Icon name={pageSVG} size="24px" />;
    }
  };

  handleClickOutside = e => {
    if (
      this.objectBrowser &&
      doesNodeContainClick(this.objectBrowser.current, e)
    )
      return;
    this.props.closeObjectBrowser();
  };

  objectBrowser = React.createRef();

  navigateTo = id => {
    this.props.searchContent(
      id,
      {
        'path.depth': 1,
        sort_on: 'getObjPositionInParent',
        metadata_fields: '_all',
      },
      `${this.props.tile}-${this.props.mode}`,
    );
    const parent = `${join(id.split('/').slice(0, -1), '/')}` || '/';
    this.setState(() => ({
      parentFolder: parent,
      currentFolder: id || '/',
    }));
  };

  toggleSearchInput = () =>
    this.setState(prevState => ({
      showSearchInput: !prevState.showSearchInput,
    }));

  onSearch = e => {
    const text = e.target.value;
    text.length > 2
      ? this.props.searchContent(
          '/',
          {
            SearchableText: `${text}*`,
            metadata_fields: '_all',
          },
          `${this.props.tile}-${this.props.mode}`,
        )
      : this.props.searchContent(
          '/',
          {
            'path.depth': 1,
            sort_on: 'getObjPositionInParent',
            metadata_fields: '_all',
          },
          `${this.props.tile}-${this.props.mode}`,
        );
  };

  onSelectItem = url => {
    if (this.props.mode === 'image') {
      this.props.onChangeTile(this.props.tile, {
        ...this.props.data,
        url: `${settings.apiPath}${url}`,
      });
      this.setState({
        selectedImage: url,
        currentImageFolder: getParentURL(url),
      });
    } else {
      this.props.onChangeTile(this.props.tile, {
        ...this.props.data,
        href: url,
      });
      this.setState({
        selectedHref: url,
        currentLinkFolder: getParentURL(url),
      });
    }
  };

  onChangeTileData = (key, value) => {
    this.props.onChangeTile(this.props.tile, {
      ...this.props.data,
      [key]: value,
    });
  };

  handleClickOnItem = item => {
    if (this.props.mode === 'image') {
      if (item.is_folderish) {
        this.navigateTo(item['@id']);
      }
      if (settings.imageObjects.includes(item['@type'])) {
        this.onSelectItem(item['@id']);
      }
    } else {
      this.onSelectItem(item['@id']);
    }
  };

  handleDoubleClickOnItem = item => {
    if (this.props.mode === 'image') {
      if (item.is_folderish) {
        this.navigateTo(item['@id']);
      }
      if (settings.imageObjects.includes(item['@type'])) {
        this.onSelectItem(item['@id']);
        this.props.closeObjectBrowser();
      }
    } else {
      this.onSelectItem(item['@id']);
      this.props.closeObjectBrowser();
    }
  };

  /**
   * Render method.
   * @method render
   * @returns {string} Markup for the component.
   */
  render() {
    return ReactDOM.createPortal(
      <aside
        role="presentation"
        onClick={e => {
          e.stopPropagation();
        }}
        ref={this.objectBrowser}
        key="objectbrowsercontainerkey"
        className="sidebar-container"
      >
        <Segment.Group raised>
          <header className="header pulled">
            <div className="vertical divider" />
            {this.state.currentFolder === '/' ? (
              <>
                {this.props.mode === 'image' ? (
                  <Icon name={folderSVG} size="24px" />
                ) : (
                  <Icon name={linkSVG} size="24px" />
                )}
              </>
            ) : (
              <Icon
                name={backSVG}
                size="24px"
                onClick={() => this.navigateTo(this.state.parentFolder)}
              />
            )}
            {this.state.showSearchInput ? (
              <form>
                <Input
                  className="search"
                  onChange={this.onSearch}
                  placeholder={this.props.intl.formatMessage(
                    messages.SearchInputPlaceholder,
                  )}
                />
              </form>
            ) : this.props.mode === 'image' ? (
              <h2>
                <FormattedMessage
                  id="ChooseImage"
                  defaultMessage="Choose Image"
                />
              </h2>
            ) : (
              <h2>
                <FormattedMessage
                  id="ChooseTargetLink"
                  defaultMessage="Choose Target"
                />
              </h2>
            )}

            <button onClick={this.toggleSearchInput}>
              <Icon name={searchSVG} size="24px" />
            </button>
            <button onClick={this.props.closeObjectBrowser}>
              <Icon name={clearSVG} size="24px" />
            </button>
          </header>
          <Segment secondary>{this.state.currentFolder}</Segment>

          <ObjectBrowserNav
            currentSearchResults={
              this.props.searchSubrequests[
                `${this.props.tile}-${this.props.mode}`
              ]
            }
            selected={
              this.props.mode === 'image'
                ? this.state.selectedImage
                : this.state.selectedHref
            }
            getIcon={this.getIcon}
            handleClickOnItem={this.handleClickOnItem}
            handleDoubleClickOnItem={this.handleDoubleClickOnItem}
            mode={this.props.mode}
            navigateTo={this.navigateTo}
          />
        </Segment.Group>
      </aside>,
      document.body,
    );
  }
}

export default compose(
  injectIntl,
  connect(
    state => ({
      searchSubrequests: state.search.subrequests,
    }),
    { searchContent },
  ),
)(ObjectBrowserBody);
