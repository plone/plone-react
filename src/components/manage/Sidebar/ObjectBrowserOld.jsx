import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { defineMessages, injectIntl, intlShape } from 'react-intl';
import { Button, Input, Segment } from 'semantic-ui-react';
import { join } from 'lodash';
import { searchContent } from '@plone/volto/actions';
import { Icon, TextWidget } from '@plone/volto/components';
import cx from 'classnames';
import { doesNodeContainClick } from 'semantic-ui-react/dist/commonjs/lib';

import { settings } from '~/config';
import backSVG from '@plone/volto/icons/back.svg';
import pageSVG from '@plone/volto/icons/page.svg';
import folderSVG from '@plone/volto/icons/folder.svg';
import clearSVG from '@plone/volto/icons/clear.svg';
import searchSVG from '@plone/volto/icons/zoom.svg';
import linkSVG from '@plone/volto/icons/link.svg';
import imageSVG from '@plone/volto/icons/image.svg';

import ObjectBrowserNav from './ObjectBrowserNav';

const messages = defineMessages({
  ImageTileInputPlaceholder: {
    id: 'Browse or type URL',
    defaultMessage: 'Browse or type URL',
  },
});

function getParentURL(url) {
  return (
    `${join(url.split('/').slice(0, -1), '/')}`.replace(settings.apiPath, '') ||
    '/'
  );
}

/**
 * ObjectBrowser container class.
 * @class ObjectBrowser
 * @extends Component
 */
@injectIntl
@connect(
  state => ({
    searchSubrequests: state.search.subrequests,
  }),
  { searchContent },
)
class ObjectBrowser extends Component {
  /**
   * Property types.
   * @property {Object} propTypes Property types.
   * @static
   */
  static propTypes = {
    tile: PropTypes.string.isRequired,
    data: PropTypes.objectOf(PropTypes.any).isRequired,
    searchSubrequests: PropTypes.objectOf(PropTypes.any).isRequired,
    searchContent: PropTypes.func.isRequired,
    closeBrowser: PropTypes.func.isRequired,
    onChangeTile: PropTypes.func.isRequired,
    intl: intlShape.isRequired,
  };

  /**
   * Default properties.
   * @property {Object} defaultProps Default properties.
   * @static
   */
  static defaultProps = {
    image: '',
    href: '',
  };

  /**
   * Constructor
   * @method constructor
   * @param {Object} props Component properties
   * @constructs WysiwygEditor
   */
  constructor(props) {
    super(props);
    this.state = {
      currentFolder: this.props.data.url
        ? getParentURL(this.props.data.url)
        : '/',
      currentImageFolder: this.props.data.url
        ? getParentURL(this.props.data.url)
        : '/',
      currentLinkFolder: this.props.data.href
        ? getParentURL(this.props.data.href)
        : '/',
      parentFolder: '',
      selectedImage: this.props.data.url
        ? this.props.data.url.replace(settings.apiPath, '')
        : '',
      selectedHref: this.props.data.href
        ? this.props.data.href.replace(settings.apiPath, '')
        : '',
      showSearchInput: false,
      alt: this.props.data.alt,
      caption: this.props.data.caption,
      external: this.props.data.external,
      mode: 'image',
    };
  }

  /**
   * Component did mount
   * @method componentDidMount
   * @returns {undefined}
   */
  componentDidMount() {
    document.addEventListener('mousedown', this.handleClickOutside, false);
    this.initialSearch('image');
  }

  /**
   * Component will receive props
   * @method componentWillUnmount
   * @returns {undefined}
   */
  componentWillUnmount() {
    document.removeEventListener('mousedown', this.handleClickOutside, false);
  }

  initialSearch = mode => {
    const currentSelected =
      mode === 'image' ? this.state.selectedImage : this.state.selectedHref;
    if (currentSelected) {
      this.props.searchContent(
        getParentURL(currentSelected),
        {
          'path.depth': 1,
          sort_on: 'getObjPositionInParent',
          metadata_fields: '_all',
        },
        `${this.props.tile}-${mode}`,
      );
    } else {
      this.props.searchContent(
        '/',
        {
          'path.depth': 1,
          sort_on: 'getObjPositionInParent',
          metadata_fields: '_all',
        },
        `${this.props.tile}-${mode}`,
      );
    }
  };

  onChangeField = (name, value) => {
    this.setState({ [name]: value });
    this.onChangeTileData(name, value);
  };

  getIcon = icon => {
    switch (icon) {
      case 'Folder':
        return <Icon name={folderSVG} size="24px" />;
      case 'Document':
        return <Icon name={pageSVG} size="24px" />;
      case 'Image':
        return <Icon name={imageSVG} size="24px" />;
      case 'File':
        return <Icon name={pageSVG} size="24px" />;
      default:
        return <Icon name={pageSVG} size="24px" />;
    }
  };

  handleClickOutside = e => {
    if (
      this.objectBrowser &&
      doesNodeContainClick(this.objectBrowser.current, e)
    )
      return;
    this.props.closeBrowser();
  };

  objectBrowser = React.createRef();

  navigateTo = id => {
    this.props.searchContent(
      id,
      {
        'path.depth': 1,
        sort_on: 'getObjPositionInParent',
        metadata_fields: '_all',
      },
      `${this.props.tile}-${this.state.mode}`,
    );
    const parent = `${join(id.split('/').slice(0, -1), '/')}` || '/';
    this.setState(() => ({
      parentFolder: parent,
      currentFolder: id || '/',
    }));
  };

  toggleSearchInput = () =>
    this.setState(prevState => ({
      showSearchInput: !prevState.showSearchInput,
    }));

  toggleMode = () => {
    if (this.state.mode === 'image') {
      this.setState(state => ({
        mode: 'link',
        currentFolder: state.currentLinkFolder,
      }));
      this.initialSearch('link');
    } else {
      this.setState(state => ({
        mode: 'image',
        currentFolder: state.currentImageFolder,
      }));
    }
  };

  onSearch = e => {
    const text = e.target.value;
    text.length > 2
      ? this.props.searchContent(
          '/',
          {
            SearchableText: `${text}*`,
            metadata_fields: '_all',
          },
          `${this.props.tile}-${this.state.mode}`,
        )
      : this.props.searchContent(
          '/',
          {
            'path.depth': 1,
            sort_on: 'getObjPositionInParent',
            metadata_fields: '_all',
          },
          `${this.props.tile}-${this.state.mode}`,
        );
  };

  onSelectItem = url => {
    if (this.state.mode === 'image') {
      this.props.onChangeTile(this.props.tile, {
        ...this.props.data,
        url: `${settings.apiPath}${url}`,
      });
      this.setState({
        selectedImage: url,
        currentImageFolder: getParentURL(url),
      });
    } else {
      this.props.onChangeTile(this.props.tile, {
        ...this.props.data,
        href: url,
      });
      this.setState({
        selectedHref: url,
        currentLinkFolder: getParentURL(url),
      });
    }
  };

  onChangeTileData = (key, value) => {
    this.props.onChangeTile(this.props.tile, {
      ...this.props.data,
      [key]: value,
    });
  };

  handleClickOnItem = item => {
    if (this.state.mode === 'image') {
      if (item.is_folderish) {
        this.navigateTo(item['@id']);
      }
      if (settings.imageObjects.includes(item['@type'])) {
        this.onSelectItem(item['@id']);
      }
    } else {
      if (!this.state.external) {
        this.onSelectItem(item['@id']);
      }
    }
  };

  handleDoubleClickOnItem = item => {
    if (this.state.mode === 'image') {
      if (item.is_folderish) {
        this.navigateTo(item['@id']);
      }
      if (settings.imageObjects.includes(item['@type'])) {
        this.onSelectItem(item['@id']);
        this.props.closeBrowser();
      }
    } else {
      if (item.is_folderish && !this.state.external) {
        this.navigateTo(item['@id']);
      }
    }
  };

  /**
   * Render method.
   * @method render
   * @returns {string} Markup for the component.
   */
  render() {
    return (
      <aside
        onClick={e => {
          e.stopPropagation();
        }}
        ref={this.objectBrowser}
        key="objectbrowser"
        className="object-browser-container"
      >
        <Segment.Group raised>
          <header className="header pulled">
            <div className="vertical divider" />
            {this.state.currentFolder === '/' ? (
              <>
                {this.state.mode === 'image' ? (
                  <Icon name={folderSVG} size="24px" />
                ) : (
                  <Icon name={linkSVG} size="24px" />
                )}
              </>
            ) : (
              <Icon
                name={backSVG}
                size="24px"
                onClick={() => this.navigateTo(this.state.parentFolder)}
              />
            )}
            {this.state.showSearchInput ? (
              <form>
                <Input
                  className="search"
                  onChange={this.onSearch}
                  placeholder={this.props.intl.formatMessage(
                    messages.ImageTileInputPlaceholder,
                  )}
                />
              </form>
            ) : this.state.mode === 'image' ? (
              <h2>Add Image</h2>
            ) : (
              <h2>Add Link</h2>
            )}

            <button onClick={this.toggleSearchInput}>
              <Icon name={searchSVG} size="24px" />
            </button>
            <button onClick={this.props.closeBrowser}>
              <Icon name={clearSVG} size="24px" color="#786Ec5D" />
            </button>
          </header>
          <Segment secondary>{this.state.currentFolder}</Segment>

          <ObjectBrowserNav
            currentSearchResults={
              this.props.searchSubrequests[
                `${this.props.tile}-${this.state.mode}`
              ]
            }
            selected={
              this.state.mode === 'image'
                ? this.state.selectedImage
                : this.state.selectedHref
            }
            getIcon={this.getIcon}
            handleClickOnItem={this.handleClickOnItem}
            handleDoubleClickOnItem={this.handleDoubleClickOnItem}
            mode={this.state.mode}
            isExternalSelected={this.state.external}
          />
        </Segment.Group>
      </aside>
    );
  }
}

export default ObjectBrowser;
