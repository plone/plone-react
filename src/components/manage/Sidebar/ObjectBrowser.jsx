import React from 'react';
import { CSSTransition } from 'react-transition-group';
import ObjectBrowserBody from './ObjectBrowserBody';

const DEFAULT_TIMEOUT = 500;

const withObjectBrowser = WrappedComponent =>
  class extends React.Component {
    constructor() {
      super();
      this.state = { isObjectBrowserOpen: false };
    }

    /**
     * openObjectBrowser
     * @function openObjectBrowser
     * @param {Object} object ObjectBrowser configuration.
     * @param {string} object.mode Quick mode, defaults to `image`.
     * @param {string} object.dataName Name of the block data property to write the selected item.
     * @param {string} object.onSelectItem Function that will be called on item selection.
     *
     * Usage:
     *
     * this.props.openObjectBrowser();
     *
     * this.props.openObjectBrowser({mode: 'link'});
     *
     * this.props.openObjectBrowser({
     *   dataName: 'myfancydatafield'
     *   });
     *
     * this.props.openObjectBrowser({
     *   onSelectItem: url =>
     *     this.props.onChangeBlock(this.props.block, {
     *       ...this.props.data,
     *       myfancydatafield: url,
     *     }),
     *   });
     */
    openObjectBrowser = ({
      mode = 'image',
      onSelectItem = null,
      dataName = null,
    } = {}) =>
      this.setState({
        isObjectBrowserOpen: true,
        mode,
        onSelectItem,
        dataName,
      });

    closeObjectBrowser = () => this.setState({ isObjectBrowserOpen: false });

    render() {
      return (
        <>
          <WrappedComponent
            {...this.props}
            isObjectBrowserOpen={this.state.isObjectBrowserOpen}
            openObjectBrowser={this.openObjectBrowser}
            closeObjectBrowser={this.closeObjectBrowser}
          />
          <CSSTransition
            in={this.state.isObjectBrowserOpen}
            timeout={DEFAULT_TIMEOUT}
            classNames="sidebar-container"
            unmountOnExit
          >
            <ObjectBrowserBody
              {...this.props}
              closeObjectBrowser={this.closeObjectBrowser}
              mode={this.state.mode}
              onSelectItem={this.state.onSelectItem}
              dataName={this.state.dataName}
            />
          </CSSTransition>
        </>
      );
    }
  };
export default withObjectBrowser;
