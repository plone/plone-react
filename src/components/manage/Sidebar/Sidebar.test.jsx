import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-intl-redux';

import Sidebar from './Sidebar';

const mockStore = configureStore();

test('renders a sidebar component', () => {
  const store = mockStore({
    intl: {
      locale: 'en',
      messages: {},
    },
    sidebar: {
      tab: 0,
    },
  });
  const component = renderer.create(
    <Provider store={store}>
      <Sidebar />
    </Provider>,
  );
  const json = component.toJSON();
  expect(json).toMatchSnapshot();
});
