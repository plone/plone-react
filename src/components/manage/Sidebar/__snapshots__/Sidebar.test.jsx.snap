// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a sidebar component 1`] = `
Array [
  <div
    className="sidebar-container"
  >
    <button
      aria-label="Shrink sidebar"
      className="ui button trigger"
      onClick={[Function]}
    />
    <div
      className="tabs-wrapper"
    >
      <div
        className="ui pointing secondary attached tabular formtabs menu"
      >
        <a
          className="item"
          onClick={[Function]}
        >
          Metadata
        </a>
        <a
          className="active item"
          onClick={[Function]}
        >
          Properties
        </a>
      </div>
      <div
        className="ui bottom attached segment tab tab-wrapper"
        id="sidebar-metadata"
      />
      <div
        className="ui bottom attached segment active tab tab-wrapper"
        id="sidebar-properties"
      >
        <svg
          className="icon tab-forbidden"
          dangerouslySetInnerHTML={
            Object {
              "__html": undefined,
            }
          }
          onClick={null}
          style={
            Object {
              "fill": "currentColor",
              "height": "48px",
              "width": "auto",
            }
          }
          viewBox=""
          xmlns=""
        />
      </div>
    </div>
  </div>,
  <div
    className="pusher expanded"
  />,
]
`;
